besu/src/main/java/org/hyperledger/besu/services/TraceServiceImpl.java

```diff
@@ -0,0 +1,119 @@
+/*
+ * Copyright Hyperledger Besu Contributors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+package org.hyperledger.besu.services;
+
+import static com.google.common.base.Preconditions.checkArgument;
+
+import org.hyperledger.besu.datatypes.DataGas;
+import org.hyperledger.besu.datatypes.Wei;
+import org.hyperledger.besu.ethereum.api.jsonrpc.internal.methods.TraceBlock.ChainUpdater;
+import org.hyperledger.besu.ethereum.api.jsonrpc.internal.processor.Tracer;
+import org.hyperledger.besu.ethereum.api.query.BlockchainQueries;
+import org.hyperledger.besu.ethereum.chain.Blockchain;
+import org.hyperledger.besu.ethereum.core.Block;
+import org.hyperledger.besu.ethereum.core.BlockHeader;
+import org.hyperledger.besu.ethereum.mainnet.MainnetTransactionProcessor;
+import org.hyperledger.besu.ethereum.mainnet.ProtocolSchedule;
+import org.hyperledger.besu.ethereum.mainnet.ProtocolSpec;
+import org.hyperledger.besu.ethereum.processing.TransactionProcessingResult;
+import org.hyperledger.besu.ethereum.vm.CachingBlockHashLookup;
+import org.hyperledger.besu.evm.tracing.OperationTracer;
+import org.hyperledger.besu.plugin.Unstable;
+import org.hyperledger.besu.plugin.services.TraceService;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/** The Trace service implementation. */
+@Unstable
+public class TraceServiceImpl implements TraceService {
+  private static final Logger LOG = LoggerFactory.getLogger(TraceServiceImpl.class);
+
+  private final BlockchainQueries blockchainQueries;
+  private final ProtocolSchedule protocolSchedule;
+
+  /**
+   * Instantiates a new TraceServiceImpl service.
+   *
+   * @param blockchainQueries the blockchainQueries
+   * @param protocolSchedule the protocolSchedule
+   */
+  public TraceServiceImpl(
+      final BlockchainQueries blockchainQueries, final ProtocolSchedule protocolSchedule) {
+    this.blockchainQueries = blockchainQueries;
+    this.protocolSchedule = protocolSchedule;
+  }
+
+  /**
+   * Traces block
+   *
+   * @param blockNumber the block number to be traced
+   * @param tracer an instance of OperationTracer
+   */
+  @Override
+  public void traceBlock(final long blockNumber, final Object tracer) {
+    checkArgument(tracer instanceof OperationTracer, "Invalid OperationTracer");
```

Comment:
Currently, to make traceBlock pass an OperationTracer directly, referencing the evm package would be necessary. However, this would result in a circular dependency:

`besu.evm` --> `besu.crypto` --> `besu.plugin` - - - > `besu.evm` 
`besu.evm` --> `besu.datatypes`  --> `besu.plugin` - - - > `besu.evm`

One possible solution is to add a plugin dependency to the EVM package and have OperationTracer implement a plugin interface. This would avoid the circular dependency issue and enable passing OperationTracer directly to traceBlock but would add a new dependency to the EVM. 

Any thoughts?

Reply:
Perhaps we should flip the plugin-api/datatypes dependency, and split crypto into algorithms and services:

```mermaid
flowchart LR;
    besu.crypto.services --> besu.crypto.algorithms;
    besu.evm-->besu.crypto.algorithms;
    besu.evm-->besu.datatypes;
    besu.plugin-api --> besu.datatypes;
```

This removes the circular dependency if we have plugin-api reference the EVM.  We may want to create a second evm-apis module possibly so the external plugins don't have to import the whole EVM.

Here's a PR - #5285

Reply:
Thank you @shemnon That solved the issue! I have updated this PR